In [14]:
import pandas as pd
from pathlib import Path
pd.set_option("display.max_rows", 50)

RUTA_BTS = Path("../data/raw/bts")
archivos = sorted(RUTA_BTS.glob("*.csv"))
for a in archivos:
    print(a.name)


# Solo la cabecera del primer mes, para ver los nombres exactos de las columnas
cabecera = pd.read_csv(archivos[0], nrows=0)
print(len(cabecera.columns), "columnas")
print(list(cabecera.columns))

On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2025_1.csv
On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2025_2.csv
On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2025_3.csv
110 columnas
['Year', 'Quarter', 'Month', 'DayofMonth', 'DayOfWeek', 'FlightDate', 'Reporting_Airline', 'DOT_ID_Reporting_Airline', 'IATA_CODE_Reporting_Airline', 'Tail_Number', 'Flight_Number_Reporting_Airline', 'OriginAirportID', 'OriginAirportSeqID', 'OriginCityMarketID', 'Origin', 'OriginCityName', 'OriginState', 'OriginStateFips', 'OriginStateName', 'OriginWac', 'DestAirportID', 'DestAirportSeqID', 'DestCityMarketID', 'Dest', 'DestCityName', 'DestState', 'DestStateFips', 'DestStateName', 'DestWac', 'CRSDepTime', 'DepTime', 'DepDelay', 'DepDelayMinutes', 'DepDel15', 'DepartureDelayGroups', 'DepTimeBlk', 'TaxiOut', 'WheelsOff', 'WheelsOn', 'TaxiIn', 'CRSArrTime', 'ArrTime', 'ArrDelay', 'ArrDelayMinutes', 'ArrDel15', 'ArrivalDelayGroups', 'ArrTimeBlk', 'Cancelled', 'Cance

In [15]:
#seleccionamos las columnas que queremos para nuestros analisis 
columnas = ["Year", "Month", "DayofMonth", "DayOfWeek", "FlightDate", "Reporting_Airline","Tail_Number", "Flight_Number_Reporting_Airline", "Origin", "OriginCityName","OriginState", "Dest", "DestCityName", "DestState", "CRSDepTime", "DepTime", "DepDelay","DepDel15", "DepTimeBlk", "TaxiOut", "TaxiIn", "CRSArrTime", "ArrTime", "ArrDelay","ArrDel15", "Cancelled", "CancellationCode", "Diverted", "AirTime", "Distance", "CarrierDelay", "WeatherDelay", "NASDelay", "SecurityDelay", "LateAircraftDelay"]
# leemos los csv 
meses_csv = []                     
for archivo in archivos:
    df_mes = pd.read_csv(archivo, usecols=columnas)   
    meses_csv.append(df_mes)                        
    print(archivo.name, df_mes.shape)          

df = pd.concat(meses_csv, ignore_index=True)           


On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2025_1.csv (539747, 35)
On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2025_2.csv (504884, 35)
On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2025_3.csv (600872, 35)


Carga de datos. Se cargaron las 35 columnas seleccionadas de los tres CSV mensuales del BTS: enero (539.747 filas), febrero (504.884) y marzo (600.872). Tras concatenarlos, el DataFrame tiene 1.645.503 filas × 35 columnas y ocupa 1,3 GB en memoria, por las 11 columnas de texto (object). Se optimizará con dtype category en la limpieza.

In [20]:
df.head(-10)

,Year,Month,DayofMonth,DayOfWeek,FlightDate,Reporting_Airline,Tail_Number,Flight_Number_Reporting_Airline,Origin,OriginCityName,...,Cancelled,CancellationCode,Diverted,AirTime,Distance,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay
0,2025,1,1,3,2025-01-01,AA,N104NN,1,JFK,"New York, NY",...,0.0,NaN,0.0,345.0,2475.0,NaN,NaN,NaN,NaN,NaN
1,2025,1,2,4,2025-01-02,AA,N110AN,1,JFK,"New York, NY",...,0.0,NaN,0.0,353.0,2475.0,NaN,NaN,NaN,NaN,NaN
2,2025,1,3,5,2025-01-03,AA,N106NN,1,JFK,"New York, NY",...,0.0,NaN,0.0,347.0,2475.0,NaN,NaN,NaN,NaN,NaN
3,2025,1,4,6,2025-01-04,AA,N117AN,1,JFK,"New York, NY",...,0.0,NaN,0.0,349.0,2475.0,NaN,NaN,NaN,NaN,NaN
4,2025,1,5,7,2025-01-05,AA,N104NN,1,JFK,"New York, NY",...,0.0,NaN,0.0,347.0,2475.0,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1645488,2025,3,18,2,2025-03-18,NK,N726NK,535,FLL,"Fort Lauderdale, FL",...,0.0,NaN,0.0,138.0,973.0,NaN,NaN,NaN,NaN,NaN
1645489,2025,3,19,3,2025-03-19,NK,N742NK,535,FLL,"Fort Lauderdale, FL",...,0.0,NaN,0.0,132.0,973.0,NaN,NaN,NaN,NaN,NaN
1645490,2025,3,20,4,2025-03-20,NK,N706NK,535,FLL,"Fort Lauderdale, FL",...,0.0,NaN,0.0,127.0,973.0,NaN,NaN,NaN,NaN,NaN
1645491,2025,3,21,5,2025-03-21,NK,N712NK,535,FLL,"Fort Lauderdale, FL",...,0.0,NaN,0.0,131.0,973.0,NaN,NaN,NaN,NaN,NaN


In [16]:
print(df.shape)
df.info(memory_usage="deep")

(1645503, 35)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1645503 entries, 0 to 1645502
Data columns (total 35 columns):
 #   Column                           Non-Null Count    Dtype  
---  ------                           --------------    -----  
 0   Year                             1645503 non-null  int64  
 1   Month                            1645503 non-null  int64  
 2   DayofMonth                       1645503 non-null  int64  
 3   DayOfWeek                        1645503 non-null  int64  
 4   FlightDate                       1645503 non-null  object 
 5   Reporting_Airline                1645503 non-null  object 
 6   Tail_Number                      1641401 non-null  object 
 7   Flight_Number_Reporting_Airline  1645503 non-null  int64  
 8   Origin                           1645503 non-null  object 
 9   OriginCityName                   1645503 non-null  object 
 10  OriginState                      1645503 non-null  object 
 11  Dest                             164

## NULOS EN EL DF

In [17]:
#veremos el porcentaje de nulos por columna 
pct_nulos = df.isna().mean() * 100
pct_nulos = pct_nulos.round(2).sort_values(ascending=False)
print(pct_nulos)

CancellationCode                   98.14
LateAircraftDelay                  80.72
SecurityDelay                      80.72
NASDelay                           80.72
WeatherDelay                       80.72
CarrierDelay                       80.72
AirTime                             2.09
ArrDel15                            2.09
ArrDelay                            2.09
ArrTime                             1.90
TaxiIn                              1.90
TaxiOut                             1.85
DepTime                             1.80
DepDelay                            1.80
DepDel15                            1.80
Tail_Number                         0.25
CRSDepTime                          0.00
DestState                           0.00
DayofMonth                          0.00
DayOfWeek                           0.00
FlightDate                          0.00
Reporting_Airline                   0.00
Distance                            0.00
Diverted                            0.00
Flight_Number_Re

## Qué dicen tus nulos 

- CancellationCode (98,14 %): solo se rellena si el vuelo se canceló, así que su nulo es la ausencia de cancelación. Por tanto, aproximadamente 100 - 98,14 = 1,86 % de vuelos cancelados.
- Las 5 columnas de causa (80,72 %): las cinco tienen exactamente el mismo porcentaje porque se rellenan a la vez y solo cuando el retraso de llegada es >15 min. El 19,28 % restante (317.266 vuelos) son los que tienen causa asignada.
- Grupo ~1,8-2,1 % (DepTime, DepDelay, ArrTime, TaxiOut, AirTime...): coincide con el orden de magnitud de las cancelaciones. Los vuelos cancelados nunca despegaron, así que no tienen horas reales.
- Las diferencias entre ellos son pistas: ArrDelay/AirTime (2,09 %) superan a DepTime (1,80 %), lo que sugiere otro grupo además de los cancelados. Mi hipótesis son los desviados (Diverted), que despegan pero no aterrizan en destino.
- Tail_Number (0,25 %): matrícula vacía, sin relación evidente con lo anterior.

## ¿CUantos vuelos hay cancelados o desviados? 
para ello primero tenemos que comprobar los valores que hay 0/1 (1 = vuelo cancelado / desviado

In [23]:
print(df["Cancelled"].unique())
print(df["Diverted"].unique())
#observamos que los valores son los correctos ahira 
##Cuantos vuelos hay cancelados? y desviados? 
mask = (df["Cancelled"] == 1) & (df["Diverted"] == 1)
print(df.loc[mask, "ArrDelay"].isna().mean())


[0. 1.]
[0. 1.]
nan


In [27]:
# 1) Cancelados: cuántos y qué porcentaje
print(df["Cancelled"].sum(), df["Cancelled"].mean() * 100)

# 2) Desviados: lo mismo
print(df["Diverted"].sum(), df["Diverted"].mean() * 100)


30640.0 1.862044614929295
3817.0 0.23196554488202087


Cancelaciones y desvíos. Hay 30.640 vuelos cancelados (1,86 %) y 3.817 desviados (0,23 %). El 1,86 % coincide con el complemento del nulo de CancellationCode (100 - 98,14), luego ese campo solo se informa en vuelos cancelados.

In [28]:
# 3) Vuelos normales (ni cancelados ni desviados)
mask = (df["Cancelled"] == 0) & (df["Diverted"] == 0)
print(mask.sum())                               # cuántas filas tiene la selección
print(df.loc[mask, "ArrDelay"].isna().mean())    # proporción de nulos en ArrDelay
print(df.loc[mask, "DepDelay"].isna().mean())    # y en DepDelay

1611046
0.0
0.0


Verificación. Hay 1.611.046 vuelos ni cancelados ni desviados (1.645.503 - 30.640 - 3.817). En ese subconjunto ArrDelay y DepDelay tienen 0 % de nulos. Por tanto, los nulos de estas dos variables proceden exclusivamente de vuelos cancelados o desviados, no de fallos de registro, y no se imputarán.

### diagnóstico de los nulos

In [29]:
# duplicados exactos (todas las columnas)
print(df.duplicated().sum())

# duplicados por identificador natural del vuelo
clave = ["FlightDate", "Reporting_Airline", "Flight_Number_Reporting_Airline", "Origin", "Dest", "CRSDepTime"]
print(df.duplicated(subset=clave).sum())

0
0


No hay filas duplicadas: df.duplicated() devuelve 0 sobre las 35 columnas, y también 0 usando como identificador natural del vuelo la combinación de fecha, aerolínea, número de vuelo, origen, destino y hora programada de salida (CRSDepTime). Por tanto, no es necesario eliminar registros y los conteos y medias posteriores no están inflados por repeticiones.

### tipos de datos

In [32]:
cols_texto = df.select_dtypes(include="object").columns
print(df[cols_texto].nunique().sort_values())
print(df["FlightDate"].head())

CancellationCode        4
Reporting_Airline      14
DepTimeBlk             19
OriginState            52
DestState              52
FlightDate             90
OriginCityName        327
DestCityName          327
Origin                333
Dest                  333
Tail_Number          5787
dtype: int64
0    2025-01-01
1    2025-01-02
2    2025-01-03
3    2025-01-04
4    2025-01-05
Name: FlightDate, dtype: object


- CancellationCode (4), Reporting_Airline (14), DepTimeBlk (19), OriginState/DestState (52): pocos valores distintos, repetidos millones de veces. Candidatas claras a category.
- OriginCityName/DestCityName (327) y Origin/Dest (333): también compensan a category (333 valores en 1,6 millones de filas es una repetición enorme).
- Tail_Number (5.787): son matrículas de aeronave. Aún repite mucho (5.787 valores en 1,6 M de filas), así que category también ahorra memoria, aunque es la menos evidente.
- FlightDate (90): no es una categoría, es una fecha. Enero (31) + febrero (28) + marzo (31) = 90 días, lo que cuadra exactamente con el periodo. Está en formato YYYY-MM-DD, así que se convierte sin ambigüedad con pd.to_datetime.

De las 35 columnas, 11 son de tipo object (texto) y explican casi todo el uso de memoria (1,3 GB). Nueve tienen pocos valores distintos respecto a las 1.645.503 filas (CancellationCode 4, Reporting_Airline 14, DepTimeBlk 19, OriginState y DestState 52, OriginCityName y DestCityName 327, Origin y Dest 333), por lo que se convertirán a category. Tail_Number (5.787 matrículas) también se beneficia de ello. FlightDate (90 valores únicos, coherentes con los 90 días de enero-marzo, formato YYYY-MM-DD) se convertirá a datetime.

### Ahorro de memoria con category

In [33]:
antes = df["Origin"].memory_usage(deep=True)
despues = df["Origin"].astype("category").memory_usage(deep=True)
print(antes / 1e6, despues / 1e6, (1 - despues / antes) * 100)

98.730308 3.31941 96.63790170694088


Como prueba, se midió la columna Origin (333 valores distintos en 1.645.503 filas): como object ocupa 98,7 MB y como category 3,3 MB, un ahorro del 96,6 %. Al aplicar la conversión a las columnas de texto de baja cardinalidad se espera una reducción muy notable de los 1,3 GB iniciales. La conversión real se realizará en el notebook de limpieza.

#### Resumen de la auditoría

Volumen. Se cargaron 35 columnas de los tres CSV mensuales del BTS (enero-marzo de 2025): 1.645.503 vuelos en total (539.747 + 504.884 + 600.872), con 1,3 GB en memoria.

Cancelaciones y desvíos. 30.640 vuelos cancelados (1,86 %) y 3.817 desviados (0,23 %). Quedan 1.611.046 vuelos ni cancelados ni desviados.

Nulos. 16 de las 35 columnas tienen nulos, y son información del proceso, no errores de carga:

CancellationCode (98,14 %) solo se informa en vuelos cancelados.
Las cinco columnas de causa de retraso (80,72 %) solo se informan cuando el retraso de llegada supera 15 minutos.
ArrDelay y DepDelay tienen 0 % de nulos entre los vuelos ni cancelados ni desviados, luego sus nulos proceden exclusivamente de cancelados y desviados.

Duplicados. No hay ninguno, ni exactos ni por identificador natural del vuelo (fecha, aerolínea, número de vuelo, origen, destino y hora programada).

Tipos. 11 columnas object concentran casi toda la memoria. Nueve tienen baja cardinalidad y se convertirán a category (Origin pasa de 98,7 MB a 3,3 MB, un 96,6 % menos); Tail_Number también se beneficia. FlightDate (90 valores, YYYY-MM-DD) se convertirá a datetime.

Pendiente para la limpieza (notebook 02).

- Verificar el origen de los nulos en TaxiOut, TaxiIn, AirTime y ArrTime.
- Revisar los 111 vuelos con DepTime pero sin DepDelay.
- Decidir qué hacer con los 4.102 nulos de Tail_Number.
- Decidir si se separan cancelados y desviados para el análisis de retrasos.
- Revisar el formato HHMM de CRSDepTime y DepTime